### PR TITLE
Feature/Inventory

### PR DESCRIPTION
**Pull request by rishabh-vinsol**

### CODE REVIEW

File: app/controllers/inventories_controller.rb [hunk 1]
```diff
@@ -0,0 +1,65 @@
+# Controller to handle Inventories CRUD
+class InventoriesController < ApplicationController
+  before_action :set_branch_id
+  before_action :set_branch, only: %i[new edit index update]
```
Comment (himanshumishra31) [inline app/controllers/inventories_controller.rb:4]: why these 2?
Only one should be enough

File: app/controllers/inventories_controller.rb [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+  private def set_branch
+    @branch = Branch.find_by(id: @branch_id)
+  end
+
+  private def get_or_set_inventory
+    if @inventory = Inventory.find_by(find_inventory_params)
+      @inventory.add_quantity(inventory_params[:quantity].to_i)
+    else
+      @inventory = Inventory.new(inventory_params)
```
Comment (himanshumishra31) [inline app/controllers/inventories_controller.rb:49]: we can handle here the scenario of adding same ingredient duplicate times

File: db/migrate/20230519094912_create_inventories.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+class CreateInventories < ActiveRecord::Migration[7.0]
+  def change
+    create_table :inventories do |t|
+      t.integer :quantity
+      t.references :branch, null: false, foreign_key: true
+      t.references :ingredient, null: false, foreign_key: true
```
Comment (himanshumishra31) [inline db/migrate/20230519094912_create_inventories.rb:6]: add unique index of `branch_id` and `ingredient_id` in database

**Review by himanshumishra31 — CHANGES_REQUESTED** [review]

File: app/views/inventories/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+<div class="container mt-4">
+  <h1 class= "mb-5"><%= link_to "Inventories", branch_inventories_path(@branch_id), class: "text-decoration-none" %> / New Inventory</h1>
```
Comment (Sparsh-Goyal-2000) [inline app/views/inventories/new.html.erb:2]: It is giving errors 
- when we aren't selecting any ingredient.
- when new ingredient is selected but quantity not selected
image

File: app/controllers/inventories_controller.rb [hunk 5]
```diff
@@ -0,0 +1,65 @@
+# Controller to handle Inventories CRUD
+class InventoriesController < ApplicationController
+  before_action :set_branch_id
+  before_action :set_branch, only: %i[new edit index update]
+  before_action :get_or_set_inventory, only: :create
+  before_action :set_inventory, only: %i[edit update destroy]
+
+  def index
+    @inventories = Inventory.includes(:ingredient).where(branch_id: @branch_id).order(:ingredient_id)
```
Comment (Sparsh-Goyal-2000) [inline app/controllers/inventories_controller.rb:9]: `branch.inventories` ...

File: app/views/branches/_branch.html.erb [hunk 6]
```diff
@@ -6,6 +6,7 @@
   <td>
     <div class= "d-flex">
       <%= link_to 'Edit', edit_branch_path(branch), class: 'btn btn-primary btn-sm me-2' %>
+      <%= link_to 'Inventory', branch_inventories_path(branch), class: 'btn btn-primary btn-sm me-2' %>
```
Comment (Sparsh-Goyal-2000) [inline app/views/branches/_branch.html.erb:9]: Rename as Manage inventory

File: app/controllers/inventories_controller.rb [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+  private def set_branch_id
+    @branch_id = params[:branch_id]
+  end
+
+  private def set_branch
+    @branch = Branch.find_by(id: @branch_id)
+  end
+
+  private def get_or_set_inventory
+    if @inventory = Inventory.find_by(find_inventory_params)
```
Comment (Sparsh-Goyal-2000) [inline app/controllers/inventories_controller.rb:46]: yes we don't need method, we can always find or initialize inventory and update its quantity

File: app/models/branch.rb [hunk 8]
```diff
@@ -4,6 +4,7 @@ class Branch < ApplicationRecord
 
   has_one :address, as: :addressable
   accepts_nested_attributes_for :address, update_only: true
+  has_many :inventories
```
Comment (Sparsh-Goyal-2000) [inline app/models/branch.rb:7]: dependent

File: app/models/ingredient.rb [hunk 9]
```diff
@@ -1,6 +1,11 @@
 # Model class ingredient
 class Ingredient < ApplicationRecord
+  ### ASSOCIATIONS ###
+
   has_one_attached :image
+  has_many :inventories
```
Comment (Sparsh-Goyal-2000) [inline app/models/ingredient.rb:6]: same

**Review by Sparsh-Goyal-2000 — CHANGES_REQUESTED** [review]

File: app/views/inventories/index.html.erb [hunk 10]
```diff
@@ -0,0 +1,28 @@
+<%= render 'shared/notice' %>
+<div class="container mt-4">
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply app/views/inventories/index.html.erb:2]: Discussed

File: db/migrate/20230519094912_create_inventories.rb [hunk 11]
```diff
@@ -0,0 +1,11 @@
+class CreateInventories < ActiveRecord::Migration[7.0]
+  def change
+    create_table :inventories do |t|
+      t.integer :quantity
+      t.references :branch, null: false, foreign_key: true
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply db/migrate/20230519094912_create_inventories.rb:5]: Default value of null for references is true. So we need to write `null: false` 
Index is present but we if we want foreign key, we have to specify

File: app/views/branches/_branch.html.erb (hunk 6 above)
Comment (rishabh-vinsol) [reply app/views/branches/_branch.html.erb:9]: Done

File: app/models/ingredient.rb (hunk 9 above)
Comment (rishabh-vinsol) [reply app/models/ingredient.rb:6]: Done

File: app/views/inventories/new.html.erb (hunk 4 above)
Comment (rishabh-vinsol) [reply app/views/inventories/new.html.erb:2]: Fixed

File: app/controllers/inventories_controller.rb [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    @inventory.destroy ? flash[:notice] = t('.success') : flash[:alert] = t('.failure')
+    redirect_to branch_inventories_path(@branch_id)
+  end
+
+  private def set_branch_id
+    @branch_id = params[:branch_id]
+  end
+
+  private def set_branch
+    @branch = Branch.find_by(id: @branch_id)
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply app/controllers/inventories_controller.rb:42]: Done

File: app/controllers/inventories_controller.rb (hunk 1 above)
Comment (rishabh-vinsol) [reply app/controllers/inventories_controller.rb:4]: Done

File: db/migrate/20230519094912_create_inventories.rb [hunk 13]
```diff
@@ -0,0 +1,11 @@
+class CreateInventories < ActiveRecord::Migration[7.0]
+  def change
+    create_table :inventories do |t|
+      t.integer :quantity
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply db/migrate/20230519094912_create_inventories.rb:4]: Done

File: app/controllers/inventories_controller.rb (hunk 2 above)
Comment (rishabh-vinsol) [reply app/controllers/inventories_controller.rb:49]: Fixed

File: app/controllers/inventories_controller.rb (hunk 7 above)
Comment (rishabh-vinsol) [reply app/controllers/inventories_controller.rb:46]: Done

File: app/views/inventories/_inventory.html.erb [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+  <td><%= inventory.ingredient.id %></td>
+  <td><%= inventory.ingredient.name %></td>
+  <td><%= inventory.quantity %></td>
+  <td>
+    <div class= "d-flex">
+      <%= link_to 'Edit', edit_branch_inventory_path(inventory.branch_id, inventory), class: 'btn btn-primary btn-sm me-2' %>
+      <%= button_to 'Delete', branch_inventory_path(inventory.branch_id, inventory), 
+        method: :delete, 
+        class: 'btn btn-danger btn-sm', 
+        data: { confirm: 'Are you sure you want to delete this branch?'}
```
Comment (Sparsh-Goyal-2000) [inline app/views/inventories/_inventory.html.erb:11]: this is not working but fix the record class
`branch` -> `inventory`

**Review by Sparsh-Goyal-2000 — APPROVED** [review]

File: app/views/inventories/_inventory.html.erb (hunk 14 above)
Comment (rishabh-vinsol) [reply app/views/inventories/_inventory.html.erb:11]: Done

**Pull request merged**